### ISSUE TITLE
peggy.js grammar for semver

### ISSUE THREAD

**Issue by dselman** [peggyjs/peggy#299]
I've recently been working on a peggy.js grammar for semver strings. I'd welcome feedback and if suitable you are welcome to add it to your library of example grammars.

https://github.com/dselman/peggy-semver/blob/main/parser.pegjs

**Comment by MarcelBolten** [comment]
Hi @dselman 

I got inspired by your work and tried to avoid the `checkNumber` function and also included some suggestions of hildjj. 
Feel free to take what you like from my attempt for your project.

~~~pegjs
semver
  = versionCore:versionCore
    pre:('-' @preRelease)?
    build:('+' @build)?
  {
    return { versionCore, pre, build };
  }

versionCore
  = major:$numericIdentifier '.' minor:$numericIdentifier '.' patch:$numericIdentifier
  {
    return {
      major: parseInt(major, 10),
      minor: parseInt(minor, 10),
      patch: parseInt(patch, 10),
    };
  }

preRelease
  = head:$preReleaseIdentifier tail:('.' @$preReleaseIdentifier)*
  {
    return [head, ...tail];
  }

build
  = head:$buildIdentifier tail:('.' @$buildIdentifier)*
  {
    return [head, ...tail];
  }

preReleaseIdentifier
  = alphanumericIdentifier
  / numericIdentifier

buildIdentifier
  = alphanumericIdentifier
  / digit+

alphanumericIdentifier
  = digit* nonDigit identifierChar*

numericIdentifier
  = '0' / (positiveDigit digit*)

identifierChar
  = [a-z0-9-]i

nonDigit
  = [a-z-]i

digit
  = [0-9]

positiveDigit
  = [1-9]
~~~

**Comment by dselman** [comment]
That's awesome - thank you @hildjj [hildjj](https://github.com/hildjj) and @MarcelBolten -- much more elegant.

**Comment by dselman** [comment]
PR created: https://github.com/peggyjs/peggy/pull/300

**Comment by hildjj** [comment]
Fixed in #304
